python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -85,14 +88,141 @@ def pyarrow_table_from_pydict(
+        if not should_serialize_as_object_ext_type:
+            raise ace
+
+        # Otherwise, attempt to fall back to serialize as python objects
+        return ArrowPythonObjectArray.from_objects(column_values)
+
+
+def _convert_to_pyarrow_native_array(
+    column_values: np.ndarray, column_name: str
 ) -> pa.Array:
+    """Converts provided NumPy `ndarray` into PyArrow's `array` while only utilizing
+    Arrow's natively supported types (ie no custom extension types)"""
+
     try:
-        return pa.array(val)
+        # NOTE: We explicitly infer PyArrow `DataType` so that
+        #       we can perform upcasting to be able to accommodate
+        #       blocks that are larger than 2Gb in size (limited
+        #       by int32 offsets used by Arrow internally)
+        dtype = _infer_pyarrow_type(column_values)
+
+        logger.debug(f"Inferred dtype of '{dtype}' for column '{column_name}'")
+
+        return pa.array(column_values, type=dtype)
     except Exception as e:
-        raise ArrowConversionError(str(enclosing_dict)) from e
+        raise ArrowConversionError(str(column_values)) from e
+
+
+def _infer_pyarrow_type(column_values: np.ndarray) -> Optional[pa.DataType]:
+    """Infers target Pyarrow `DataType` based on the provided
+    columnar values.
+
+    NOTE: This is a wrapper on top of `pa.infer_type(...)` utility
+          performing up-casting of `binary` and `string` types to
+          corresponding `large_binary` and `large_string` types in case
+          any of the array elements exceeds 2Gb in size therefore
+          making it impossible for original types to accommodate such
+          values.
+
+          Unfortunately, for unknown reasons PA doesn't perform
+          that upcasting itself henceforth we have to do perform
+          it manually
+
+    Args:
+        column_values: List of columnar values
+
+    Returns:
+        Instance of PyArrow's `DataType` based on the provided
+        column values
+    """
+
+    if len(column_values) == 0:
+        return None
+
+    inferred_pa_dtype = pa.infer_type(column_values)
+
+    def _lt_2gb(obj: Any) -> bool:
+        # NOTE: This utility could be seeing objects other than strings or bytes in
+        #       cases when column contains non-scalar non-homogeneous object types as
+        #       column values, therefore making Arrow unable to infer corresponding
+        #       column type appropriately, therefore falling back to assume the type
+        #       of the first element in the list.
+        #
+        #       Check out test cases for this method for an additional context.
+        if isinstance(obj, (str, bytes)):
+            return len(obj) > 2 * GiB
+
+        return False
+
+    if pa.types.is_binary(inferred_pa_dtype) and any(
+        [_lt_2gb(v) for v in column_values]
+    ):
+        return pa.large_binary()
+    elif pa.types.is_string(inferred_pa_dtype) and any(
+        [_lt_2gb(v) for v in column_values]
+    ):
```

Comment:
did you want to support large_lists as well

Reply:
So there are 2 cases here:

1. Cumulative byte size of the array of objects exceeding 2Gb 
2. # of objects in the array exceeding 2^31-1

First one is being handled by Arrow internally when you call `pa.array(...)` by chunking your input in chunks < 2Gb.

Second case requires actually large_lists being used and we can support that but i don't think a single Python process will survive that and more importantly i won't be even write a test for it (Python struggles even with 2Gb strings chunked into 100Mb chunks, forget 4B objects).

Reply:
Consider making the 2GB threshold an argument so that you can unit-test these without worrying about OOM.